stdlib/src/collections/list.mojo

```diff
@@ -930,6 +930,35 @@ struct List[T: CollectionElement, hint_trivial_type: Bool = False](
         """
         return self.data
 
+    fn map(ref self, func: fn (mut T) -> None) -> Self:
```

Comment:
**thought**: The API you're describing seems more like `for_each` than `map`.
`fn(mut T) -> None` is a common `for_each` signature, whereas `map` tends to be `fn[U: CollectionElement](T) -> U`.

Examples:
* Rust [for_each](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.for_each) and [map](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.map)
* Swift [for_each](https://developer.apple.com/documentation/swift/array/foreach(_:)) and [map](https://developer.apple.com/documentation/swift/sequence/map(_:))
* C++ [ranges::for_each](https://en.cppreference.com/w/cpp/algorithm/ranges/for_each) and [ranges::transform](https://en.cppreference.com/w/cpp/algorithm/ranges/transform)

Reply:
Yes I think it was just a mistake in assuming the output will be the same type as the input collection type

Reply:
Not exactly. `for_each` tends to mutate in place, while `map` tends to create a new allocation. Users associate both APIs with those characteristics in my opinion.

Reply:
I meant the OP meant map as in map a func to return a new instance with copied values changed, just that he forgot the use case where the collection type of the result might be different.

personal note: I'd actually like us to call `for_each`  `apply` because it's a name used in the pandas and numpy ecosystem and is IMO much more intuitive (see PR #3877)